File: common/src/main/java/com/zorroa/common/elastic/CountingBulkListener.java
package com.zorroa.common.elastic;

import org.elasticsearch.action.bulk.BulkProcessor;
import org.elasticsearch.action.bulk.BulkRequest;
import org.elasticsearch.action.bulk.BulkResponse;

import java.util.concurrent.atomic.LongAdder;

/**
 * Created by chambers on 6/22/17.
 */
public class CountingBulkListener implements  BulkProcessor.Listener {

    private final LongAdder successCount = new LongAdder();
    private final LongAdder errorCount = new LongAdder();

    @Override
    public void beforeBulk(long l, BulkRequest bulkRequest) {

    }

    @Override
    public void afterBulk(long l, BulkRequest bulkRequest, BulkResponse bulkResponse) {
        successCount.add(bulkResponse.getItems().length);
    }

    @Override
    public void afterBulk(long l, BulkRequest bulkRequest, Throwable throwable) {
        errorCount.increment();
    }

    public long getSuccessCount() {
        return successCount.longValue();
    }

    public long getErrorCount() {
        return errorCount.longValue();
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/elastic/CountingBulkListener.kt
package com.zorroa.archivist.elastic

import org.elasticsearch.action.bulk.BulkProcessor
import org.elasticsearch.action.bulk.BulkRequest
import org.elasticsearch.action.bulk.BulkResponse
import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.LongAdder

/**
 * Created by chambers on 6/22/17.
 */
class CountingBulkListener : BulkProcessor.Listener {

    val successCount = LongAdder()
    val errorCount = LongAdder()

    override fun beforeBulk(l: Long, bulkRequest: BulkRequest) {

    }

    override fun afterBulk(l: Long, bulkRequest: BulkRequest, bulkResponse: BulkResponse) {
        successCount.add(bulkResponse.items.size.toLong())
    }

    override fun afterBulk(l: Long, bulkRequest: BulkRequest, throwable: Throwable) {
        errorCount.increment()
        logger.warn("Failed to process bulk request: ", throwable)
    }

    fun getSuccessCount(): Long {
        return successCount.toLong()
    }

    fun getErrorCount(): Long {
        return errorCount.toLong()
    }

    companion object {

        private val logger = LoggerFactory.getLogger(CountingBulkListener::class.java)
    }
}
